src/datasource.ts

```diff
@@ -16,23 +16,107 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import { DataSourceInstanceSettings, ScopedVars } from '@grafana/data';
-import { DataSourceWithBackend, getTemplateSrv } from '@grafana/runtime';
+import { DataFrame, DataSourceInstanceSettings, ScopedVars, TimeRange, toDataFrame } from '@grafana/data';
+import { BackendSrv, DataSourceWithBackend, FetchResponse, getBackendSrv, getTemplateSrv } from '@grafana/runtime';
 import { PixieDataSourceOptions, PixieDataQuery } from './types';
 
+interface ClusterMeta {
+  id: string;
+  name: string;
+}
+
 export class DataSource extends DataSourceWithBackend<PixieDataQuery, PixieDataSourceOptions> {
+  backendServ?: BackendSrv;
+
   constructor(instanceSettings: DataSourceInstanceSettings<PixieDataSourceOptions>) {
     super(instanceSettings);
+    this.backendServ = getBackendSrv();
   }
 
   applyTemplateVariables(query: PixieDataQuery, scopedVars: ScopedVars) {
+    const pxlScript = query?.queryBody?.pxlScript;
     return {
       ...query,
-      pxlScript: query.pxlScript
-        ? getTemplateSrv().replace(query.pxlScript, {
-            ...scopedVars,
-          })
-        : '',
+      queryBody: {
+        pxlScript: pxlScript
+          ? getTemplateSrv().replace(pxlScript, {
+              ...scopedVars,
+            })
+          : '',
+      },
+    };
+  }
+
+  async fetchMetricNames(options: any): Promise<FetchResponse | void> {
+    let refId = 'tempvar';
+    if (options && options.variable && options.variable.name) {
+      refId = options.variable.name;
+    }
+
+    const range = options?.range as TimeRange;
+    const interpolatedQuery = {
+      refId: refId,
+      datasource: {
+        type: this.type,
+        uid: this.uid,
+      },
+      queryType: 'get-clusters',
+    };
+
+    options = {
+      ...options,
+      url: '/api/ds/query',
+      method: 'POST',
+      data: {
+        from: range?.from?.valueOf()?.toString(),
+        to: range?.to?.valueOf()?.toString(),
+        queries: [interpolatedQuery],
+      },
     };
+
+    const response = this.backendServ?.fetch<any>(options);
+    if (response) {
+      return response.toPromise();
+    } else {
+      return Promise.resolve();
+    }
+  }
+
+  /**
+   * Zip fields in the dataframe into a list of individual objects with fields from dataframe.
+   * Ex: if df has n entries, and fields `a` and `b`, this function
+   * will return [{a:value of a in first row of df, b: value of b in the first row of df},
+   *  *values from the second row*, ...]
+   *
+   * @param df grafana dataframe
+   * @returns list of objects with fields names from df
+   */
+  zipGrafanaDataFrame(df: DataFrame) {
+    const zippedData = new Array(df.length);
+
+    for (let i = 0; i < df.length; i++) {
+      zippedData[i] = {};
+    }
+
+    df.fields.forEach((field) => {
+      field.values.toArray().forEach((elem, i) => {
+        zippedData[i][field.name] = elem;
+      });
+    });
+
+    console.log(zippedData);
+    return zippedData;
```

Comment:
Leftover log, and you don't need to call so many functions:

```suggestion
    const zipped = Array(df.length).fill(null);

    for (const field of df.fields) {
      for (let i = 0; i < field.values.length; i++) {
        zipped[i] ??= {};
        zipped[i][field.name] = field.values.get(i);
      }
    }

    return zipped;
```

Reply:
Thanks, this is definitely a more readable function implementation.